pilot/pkg/networking/core/v1alpha3/envoyfilter/rollup.go

```diff
@@ -0,0 +1,130 @@
+	networking.EnvoyFilter_HTTP_FILTER:    doHttpFilterMerge,
+}
+
+func rollupPatches(
+	patchContext networking.EnvoyFilter_PatchContext,
+	allPatches map[networking.EnvoyFilter_ApplyTo][]*model.EnvoyFilterConfigPatchWrapper,
+) map[networking.EnvoyFilter_ApplyTo][]*model.EnvoyFilterConfigPatchWrapper {
+	rollup := make(map[networking.EnvoyFilter_ApplyTo][]*model.EnvoyFilterConfigPatchWrapper, len(allPatches))
+	for applyTo, patches := range allPatches {
+		if _, ok := mergeFuncByType[applyTo]; !ok {
+			continue
+		}
+		if len(patches) < 2 {
+			continue
+		}
+
+		outPatches := make([]*model.EnvoyFilterConfigPatchWrapper, 0, len(patches))
+		byMatch := map[string]*model.EnvoyFilterConfigPatchWrapper{}
+
+		for _, cp := range patches {
+			if !commonConditionMatch(patchContext, cp) {
+				continue
+			}
+			if cp.Operation != networking.EnvoyFilter_Patch_MERGE {
+				outPatches = append(outPatches, cp)
+				continue
+			}
+
+			var groupKey string
+			if globalMatch(cp) {
+				groupKey = "global"
+			} else {
+				var err error
+				if groupKey, err = matcherHash(cp); err != nil {
+					// can't perform group by, but we don't want to lose the patch
+					outPatches = append(outPatches, cp)
+					continue
+				}
+			}
+
+			rollupPatch, ok := byMatch[groupKey]
+			if ok {
+				mergeFuncByType[applyTo](rollupPatch.Value, cp)
+			} else {
+				byMatch[groupKey] = newRollupPatch(cp)
+			}
+		}
+		for _, rollupPatch := range byMatch {
+			outPatches = append(outPatches, rollupPatch)
+		}
+		rollup[applyTo] = outPatches
+	}
+	return rollup
+}
+
+func globalMatch(cp *model.EnvoyFilterConfigPatchWrapper) bool {
+	switch cp.ApplyTo {
+	case networking.EnvoyFilter_CLUSTER:
+		return clusterMatch(nil, cp)
+	case networking.EnvoyFilter_LISTENER:
+		return listenerMatch(nil, cp)
+	case networking.EnvoyFilter_FILTER_CHAIN:
+		return listenerMatch(nil, cp) && filterChainMatch(nil, cp)
+	case networking.EnvoyFilter_NETWORK_FILTER:
+		return listenerMatch(nil, cp) && filterChainMatch(nil, cp) &&
+			networkFilterMatch(nil, cp)
+	case networking.EnvoyFilter_HTTP_FILTER:
+		return listenerMatch(nil, cp) && filterChainMatch(nil, cp) &&
+			networkFilterMatch(nil, cp) && httpFilterMatch(nil, cp)
+	}
+	return false
+}
+
+func matcherHash(cp *model.EnvoyFilterConfigPatchWrapper) (string, error) {
+	var objectMatcher proto.Message
+	switch cp.Match.ObjectTypes.(type) {
+	case *networking.EnvoyFilter_EnvoyConfigObjectMatch_Listener:
+		objectMatcher = cp.Match.ObjectTypes.(*networking.EnvoyFilter_EnvoyConfigObjectMatch_Listener).Listener
+	case *networking.EnvoyFilter_EnvoyConfigObjectMatch_Cluster:
+		objectMatcher = cp.Match.ObjectTypes.(*networking.EnvoyFilter_EnvoyConfigObjectMatch_Cluster).Cluster
+	default:
+		return "", fmt.Errorf("unsupported object matcher in patch wrapper; got %T", cp.Match.ObjectTypes)
+	}
+
+	if objectMatcher == nil {
+		return "", fmt.Errorf("empty matcher for patch")
+	}
+
+	hash, err := proto.Marshal(objectMatcher)
```

Comment:
This is the fastest way I've found to make a key for these. I thought [deepmind/objecthash-proto](https://github.com/deepmind/objecthash-proto) might work but it's slower and has issues with 1.20+